### PR TITLE
Converters refactor [transformers-utils]: Generalize model utilities for broader transformer support

### PR DESCRIPTION
**Pull request by markurtz**
## **Summary**

Refactors transformer utilities from Eagle-specific implementations to comprehensive, generalized utilities. Extracts and enhances existing functionality into a dedicated `transformers_utils` module that supports multiple weight formats, authentication, caching, and robust error handling while maintaining backward compatibility.

## **Details**

- **Removed** Eagle-specific utility functions from `src/speculators/convert/eagle/utils.py` including:
    - `download_checkpoint_from_hub()`
    - `ensure_checkpoint_is_local()`
    - `load_checkpoint_config()`
    - `load_checkpoint_weights()`
    - `detect_fusion_bias_and_layernorms()`
- **Added** comprehensive transformers_utils.py module with enhanced utilities:
    - `download_model_checkpoint_from_hub()` - Enhanced downloading with authentication, caching, and revision support
    - `check_download_model_checkpoint()` - Smart path/model instance handling with automatic downloads
    - `check_download_model_config()` - Flexible config loading from various sources
    - `load_model_config()` - Unified config loading with AutoConfig integration
    - `load_model_checkpoint_config_dict()` - Dictionary-based config loading
    - `load_model_checkpoint_weight_files()` - Multi-format weight file discovery
    - `load_model_checkpoint_index_weight_files()` - Indexed/sharded weight file support
    - `load_model_checkpoint_state_dict()` - Complete state dictionary loading with format detection
- **Enhanced features** compared to original Eagle utilities:
    - Support for both SafeTensors (.safetensors) and PyTorch (.bin) weight formats
    - Automatic format detection and sharded checkpoint support
    - Authentication token support for private models
    - Comprehensive type hints and modern Python syntax (union types, Path objects)
    - Robust error handling with detailed exception messages
    - Model instance support (PreTrainedModel, nn.Module) alongside path-based loading
- **Updated** __init__.py to export all new transformer utility functions
- **Replaced** Eagle-specific tests with comprehensive test suite for `transformers_utils`:
    - Enhanced test coverage with smoke, sanity, and regression test categories
    - Parameterized testing for multiple input types and edge cases

## **Test Plan**

- Expanded and added unit tests to for the full functionality

## **Related Issues**

N/A

### CODE REVIEW

File: tests/unit/utils/test_transformers_utils.py [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+    def valid_instances(
+        self,
+        mock_pretrained_factory: tuple[
+            MockPretrainedTransformersFactory, dict[str, MockType]
+        ],
+    ) -> tuple[PretrainedBundle, MockType]:
+        # Register default mock model
+        factory, patched = mock_pretrained_factory
+
+        return factory.register(), patched[
```
Comment (rahul-tuli) [inline tests/unit/utils/test_transformers_utils.py:44]: nice!

**Review by rahul-tuli — APPROVED** [review]
